ffi/diplomat/src/bidi.rs

```diff
@@ -0,0 +1,165 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[diplomat::bridge]
+pub mod ffi {
+    use alloc::boxed::Box;
+    use diplomat_runtime::{DiplomatResult, DiplomatWriteable};
+
+    use core::fmt::Write;
+    use icu_properties::bidi::BidiClassAdapter;
+    use icu_properties::maps;
+    use icu_properties::provider::BidiClassV1Marker;
+    use icu_provider::DataPayload;
+    use unicode_bidi::BidiInfo;
+    use unicode_bidi::Level;
+    use unicode_bidi::Paragraph;
+
+    use crate::provider::ffi::ICU4XDataProvider;
+
+    pub enum ICU4XBidiDirection {
+        Ltr,
+        Rtl,
+        Mixed,
+    }
+
+    #[diplomat::opaque]
+    /// An ICU4X Bidi object, containing loaded bidi data
+    #[diplomat::rust_link(icu::properties::bidi::BidiClassAdapter, Struct)]
+    // #[diplomat::rust_link(icu::properties::maps::get_bidi_class, Struct)]
+    pub struct ICU4XBidi(pub DataPayload<BidiClassV1Marker>);
+
+    impl ICU4XBidi {
+        /// Creates a new [`ICU4XBidi`] from locale data.
+        #[diplomat::rust_link(icu::properties::bidi::BidiClassAdapter::new, FnInStruct)]
+        pub fn try_new(provider: &ICU4XDataProvider) -> DiplomatResult<Box<ICU4XBidi>, ()> {
+            use icu_provider::serde::AsDeserializingBufferProvider;
+            let provider = provider.0.as_deserializing();
+            if let Result::Ok(bidi) = maps::get_bidi_class(&provider) {
+                Ok(Box::new(ICU4XBidi(bidi))).into()
+            } else {
+                Err(()).into()
+            }
+        }
+
+        /// Use the data loaded in this object to process a string and calculate bidi information
+        #[diplomat::rust_link(unicode_bidi::BidiInfo::new_with_data_source, FnInStruct)]
+        pub fn for_text<'text>(&self, text: &'text str) -> Box<ICU4XBidiInfo<'text>> {
+            let data = self.0.get();
+            let adapter = BidiClassAdapter::new(&data.code_point_trie);
+
+            Box::new(ICU4XBidiInfo(BidiInfo::new_with_data_source(
+                &adapter, text, None,
+            )))
+        }
+    }
+
+    /// An object containing bidi information for a given string, produced by `for_text()` on `ICU4XBidi`
+    #[diplomat::rust_link(unicode_bidi::BidiInfo, Struct)]
+    #[diplomat::opaque]
+    pub struct ICU4XBidiInfo<'text>(pub BidiInfo<'text>);
+
+    impl<'text> ICU4XBidiInfo<'text> {
+        //. The number of paragraphs contained here
+        pub fn paragraph_count(&self) -> usize {
+            self.0.paragraphs.len()
+        }
+
+        /// Get the nth paragraph, returning None if out of bounds
+        pub fn paragraph_at(&'text self, n: usize) -> Option<Box<ICU4XBidiParagraph<'text>>> {
+            self.0
+                .paragraphs
+                .get(n)
+                .map(|p| Box::new(ICU4XBidiParagraph(Paragraph::new(&self.0, p))))
```

Comment:
Suggestion: how many paragraphs are there, usually? I feel like this is a hot path, and it's sad that we have to box stuff on a hot path.

I guess another way to do this would be to move all the functions from `ICU4XBidiParagraph` onto `ICU4XBidiInto` and give them an extra argument, the paragraph number. That's not great ergonomics.

Does or will Diplomat support non-opaque structs such as

```rust
pub struct ICU4XBidiParagraphV2<'info, 'text> {
    pub num: usize,
    pub info: &'info ICU4XBidiInfo<'text>
}
```

I'd even be okay moving the ownership of the BidiInfo, since it's fairly easy to get back

```rust
pub struct ICU4XBidiParagraphV2<'text> {
    pub num: usize,
    pub info: ICU4XBidiInfo<'text>
}
```

Another idea: just add the paragraph number to the data model of `ICU4XBidiInfo` itself, and make it into a mutable iterator, more like how `LineBreakIterator` works. Maybe rename the class `ICU4XBidiIterator` or something like that.

Reply:
> Does or will Diplomat support non-opaque structs such as

Diplomat has _some_ support for this, but it's harder to get methods/etc on them because you can't take them as `&self`

> I'd even be okay moving the ownership of the BidiInfo, since it's fairly easy to get back

This isn't safe in FFI due to GCs, we can't have opaque types passed around as `self`.

-----

What if we just made ICU4XBidiParagraph support a `set_paragraph_index()` function that lets you set the paragraph index within bidiinfo?

Reply:
> What if we just made ICU4XBidiParagraph support a `set_paragraph_index()` function that lets you set the paragraph index within bidiinfo?

Right. Also include a `.next()` and `.prev()` function to increment or decrement the paragraph index by 1.

I think we don't really need two types, then; we can just have one. Do you agree?

Reply:
I tried that approach, it seems pretty nice. A _bit_ strange but not too out of the way.

Reply:
You still have two classes; did you mean to consolidate them?